presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperatorFactory.java

```diff
@@ -67,29 +68,30 @@ public LookupJoinOperatorFactory(int operatorId,
         this.joinType = requireNonNull(joinType, "joinType is null");
         this.joinProbeFactory = requireNonNull(joinProbeFactory, "joinProbeFactory is null");
 
-        this.referenceCount = new ReferenceCount();
+        probeReferenceCount = new ReferenceCount();
```

Comment:
I don't like this logic here, it's getting more and more complicated and more and more implicit what depends on what. It was hard for me to understand what is going on here.

Maybe create some explicit `LookupSourceState` class, that would internally have a following states: `WAITING_FOR_HBO_TO_FINISH`, `WAITING_FOR_PROBE_TO_FINISH`, `WAITING_FOR_OUTER_OPERATORS_TO_FINISH` (with different names) and some explicit state transitions behind some nicely named methods? 

Instead of executing in `LookupOuterOperator` lines like `referenceCount.release();` you would have `lookupSourceState.outerOperatorFinished()`, etc. This shouldn't be a big change. Especially that for now, you could drop `WAITING_FOR_HBO_TO_FINISH` state from the first implementation.

Maybe this state could/should be integrated/inlined with `LookupSourceFactory` itself? Or maybe there should some wrapper around abstract `LookupSourceFactory` that adds statefulness to it?